core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java

```diff
@@ -0,0 +1,349 @@
+          && event.getClickedPlayer() != null) {
+
+        event.setCancelled(true);
+
+        freeze.setFrozen(
+            event.getPlayer().getBukkit(),
+            event.getClickedPlayer(),
+            !freeze.isFrozen(event.getClickedEntity()));
+      }
+    }
+  }
+
+  private static final List<String> ALLOWED_CMDS = Lists.newArrayList("/msg", "/r");
+
+  @EventHandler(priority = EventPriority.HIGH)
+  public void onPlayerCommand(final PlayerCommandPreprocessEvent event) {
+    if (freeze.isFrozen(event.getPlayer()) && !event.getPlayer().hasPermission(Permissions.STAFF)) {
+      boolean allow =
+          ALLOWED_CMDS.stream()
+              .filter(cmd -> event.getMessage().startsWith(cmd))
+              .findAny()
+              .isPresent();
+
+      if (!allow) {
+        // Don't allow commands except for those related to chat.
+        event.setCancelled(true);
+      }
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onPlayerMove(final PlayerMoveEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      Location old = event.getFrom();
+      old.setPitch(event.getTo().getPitch());
+      old.setYaw(event.getTo().getYaw());
+      event.setTo(old);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
+  public void onVehicleMove(final VehicleMoveEvent event) {
+    if (!event.getVehicle().isEmpty() && freeze.isFrozen(event.getVehicle().getPassenger())) {
+      event.getVehicle().setVelocity(new Vector(0, 0, 0));
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onVehicleEnter(final VehicleEnterEvent event) {
+    if (freeze.isFrozen(event.getEntered())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onVehicleExit(final VehicleExitEvent event) {
+    if (freeze.isFrozen(event.getExited())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBlockBreak(final BlockBreakEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBlockPlace(final BlockPlaceEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBucketFill(final PlayerBucketFillEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBucketEmpty(final PlayerBucketEmptyEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
```

Comment:
I believe you can just cancel PlayerBucketEvent instead of both fill and empty?

Reply:
Good idea! I’ll test that out and let you know.

Reply:
Ok it’s really weird but the `PlayerBucketEvent` caused a lot of headaches for an hour. For some reason it caused an issue where players were unable to join the server at all. So I’m just going to keep the separate events.

Reply:
Well, thanks for trying 👍